driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/pipe.bpf.c

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (C) 2022 The Falco Authors.
+ *
+ * This file is dual licensed under either the MIT or GPL 2. See MIT.txt
+ * or GPL2.txt for full copies of the license.
+ */
+
+/* These BPF programs are used both for `pipe` and `pipe2` syscalls. */
+
+#include <helpers/interfaces/fixed_size_event.h>
+
+/*=============================== ENTER EVENT ===========================*/
+
+SEC("tp_btf/sys_enter")
+int BPF_PROG(pipe_e,
+	     struct pt_regs *regs,
+	     long id)
+{
+	struct ringbuf_struct ringbuf;
+	if(!ringbuf__reserve_space(&ringbuf, PIPE_E_SIZE))
+	{
+		return 0;
+	}
+
+	ringbuf__store_event_header(&ringbuf, PPME_SYSCALL_PIPE_E, PIPE_E_SIZE);
+
+	/*=============================== COLLECT PARAMETERS  ===========================*/
+
+	// Here we have no parameters to collect.
+
+	/*=============================== COLLECT PARAMETERS  ===========================*/
+
+	ringbuf__submit_event(&ringbuf);
+
+	return 0;
+}
+
+/*=============================== ENTER EVENT ===========================*/
+
+/*=============================== EXIT EVENT ===========================*/
+
+SEC("tp_btf/sys_exit")
+int BPF_PROG(pipe_x,
+	     struct pt_regs *regs,
+	     long ret)
+{
+	struct ringbuf_struct ringbuf;
+	if(!ringbuf__reserve_space(&ringbuf, PIPE_X_SIZE))
+	{
+		return 0;
+	}
+
+	ringbuf__store_event_header(&ringbuf, PPME_SYSCALL_PIPE_X, PIPE_X_SIZE);
+
+	/*=============================== COLLECT PARAMETERS  ===========================*/
+
+	/* Parameter 1: res (type: PT_ERRNO) */
+	ringbuf__store_s64(&ringbuf, ret);
+
+	s32 pipefd[2] = {0, 0};
```

Comment:
minor: was the intention to initialize w/ zero? Zero could be valid fd, so maybe initialize with `-1`?

Reply:
that's a good point, `-1` is better than `0` thank you!

Reply:
your welcome... (I guess with the change to `-1`, the failure test case probably needs a resp. change as well).

Reply:
Another thing that i didn't know is that `bpf_probe_read_user` changes the content of the memory filling it with `0` in case of failure

```c
bpf_printk("pipe0 before: %d", pipefd[0]);
bpf_printk("pipe1 before: %d", pipefd[1]);

unsigned long fd_vector_pointer = extract__syscall_argument(regs, 0);
if(bpf_probe_read_user((void *)pipefd, sizeof(pipefd), (void *)fd_vector_pointer)!=0)
{
bpf_printk("error!");
};

bpf_printk("pipe0: %d", pipefd[0]);
bpf_printk("pipe1: %d", pipefd[1]);
```

**OUTPUT:**

```
bpf_test-18134   [011] d... 16021.222626: bpf_trace_printk: pipe0 before: -1
bpf_test-18134   [011] d... 16021.222640: bpf_trace_printk: pipe1 before: -1
bpf_test-18134   [011] d... 16021.222646: bpf_trace_printk: error!
bpf_test-18134   [011] d... 16021.222647: bpf_trace_printk: pipe0: 0
bpf_test-18134   [011] d... 16021.222648: bpf_trace_printk: pipe1: 0
```

So probably the best thing is to reinforce `-1` as value when `bpf_probe_read_user` fails:

```c
if(bpf_probe_read_user((void *)pipefd, sizeof(pipefd), (void *)fd_vector_pointer)!=0)
{
        pipefd[0] = -1;
        pipefd[1] = -1;
};
```